File: back/src/test/java/com/back/domain/ticket/controller/TicketControllerTest.java
package com.back.domain.ticket.controller;

import com.back.domain.concert.entity.Concert;
import com.back.domain.concert.repository.ConcertRepository;
import com.back.domain.schedule.entity.Schedule;
import com.back.domain.schedule.entity.ScheduleSeat;
import com.back.domain.schedule.repository.ScheduleRepository;
import com.back.domain.schedule.repository.ScheduleSeatRepository;
import com.back.domain.ticket.entity.Ticket;
import com.back.domain.ticket.repository.TicketRepository;
import com.back.domain.user.entity.LoginType;
import com.back.domain.user.entity.User;
import com.back.domain.user.repository.UserRepository;
import com.back.domain.venue.entity.Venue;
import com.back.domain.venue.repository.VenueRepository;
import com.back.global.security.SecurityUser;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.data.redis.core.HashOperations;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.data.redis.core.ValueOperations;
import org.springframework.data.redis.core.ZSetOperations;
import org.springframework.http.MediaType;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

import static com.back.domain.schedule.entity.SeatStatus.AVAILABLE;
import static com.back.domain.schedule.entity.SeatStatus.SOLD_OUT;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.user;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@ActiveProfiles("test")
@SpringBootTest
@AutoConfigureMockMvc
@Transactional
class TicketControllerTest {
    private final MockMvc mockMvc;
    private final UserRepository userRepository;
    private final ConcertRepository concertRepository;
    private final VenueRepository venueRepository;
    private final ScheduleRepository scheduleRepository;
    private final ScheduleSeatRepository scheduleSeatRepository;
    private final TicketRepository ticketRepository;

    private User user;
    private SecurityUser securityUser;
    private Concert concert;
    private Schedule schedule;
    private ScheduleSeat seat;

    @MockitoBean
    private StringRedisTemplate redisTemplate;

    @Autowired
    TicketControllerTest(
            MockMvc mockMvc,
            UserRepository userRepository,
            ConcertRepository concertRepository,
            VenueRepository venueRepository,
            ScheduleRepository scheduleRepository,
            ScheduleSeatRepository scheduleSeatRepository,
            TicketRepository ticketRepository
    ) {
        this.mockMvc = mockMvc;
        this.userRepository = userRepository;
        this.concertRepository = concertRepository;
        this.venueRepository = venueRepository;
        this.scheduleRepository = scheduleRepository;
        this.scheduleSeatRepository = scheduleSeatRepository;
        this.ticketRepository = ticketRepository;
    }

    @BeforeEach
    void setUp() {
        user = saveUser();
        securityUser = new SecurityUser(user.getUserId(), user.getName());
        ValueOperations<String, String> valueOperations = mock(ValueOperations.class);
        when(redisTemplate.opsForValue()).thenReturn(valueOperations);
        when(valueOperations.get(anyString())).thenReturn("test-queue-token");
        concert = concertRepository.save(Concert.create(
                "싸이 콘서트",
                "설명",
                LocalDateTime.now(),
                LocalDateTime.now().plusDays(1),
                "poster.jpg"
        ));
        Venue venue = venueRepository.save(Venue.create("공연장", "서울", 15000L));
        schedule = scheduleRepository.save(Schedule.create(concert, venue, LocalDateTime.now().plusHours(12), 1));

        seat = scheduleSeatRepository.save(ScheduleSeat.create(schedule, "VIP", "A-1", 150000, AVAILABLE));
        scheduleSeatRepository.save(ScheduleSeat.create(schedule, "VIP", "A-2", 150000, AVAILABLE));
        scheduleSeatRepository.save(ScheduleSeat.create(schedule, "VIP", "A-3", 150000, AVAILABLE));
        scheduleSeatRepository.save(ScheduleSeat.create(schedule, "VIP", "A-4", 150000, AVAILABLE));

        @SuppressWarnings("unchecked")
        HashOperations<String, Object, Object> hashOperations = mock(HashOperations.class);
        when(redisTemplate.opsForHash()).thenReturn(hashOperations);

        when(hashOperations.multiGet(anyString(), anyList()))
                .thenReturn(List.of(user.getUserId().toString(), "test-token"));

        @SuppressWarnings("unchecked")
        ZSetOperations<String, String> zSetOperations = mock(ZSetOperations.class);
        when(redisTemplate.opsForZSet()).thenReturn(zSetOperations);
        when(zSetOperations.score(anyString(), anyString()))
                .thenReturn((double) (System.currentTimeMillis() + 600000));
    }

    @Test
    @DisplayName("티켓 2매 생성 성공")
    void createTicket() throws Exception {
        when(redisTemplate.opsForHash().multiGet(anyString(), anyList()))
                .thenAnswer(invocation -> {
                    String key = invocation.getArgument(0);
                    if (key.endsWith("A-1")) {
                        return List.of(user.getUserId().toString(), "token-1");
                    } else if (key.endsWith("A-2")) {
                        return List.of(user.getUserId().toString(), "token-2");
                    }
                    return List.of(user.getUserId().toString(), "test-token");
                });

        String requestBody = """
                {
                  "concertId": %d,
                  "seatHolds": [
                    {
                      "seatNumber": "A-1",
                      "occupyToken": "token-1"
                    },
                    {
                      "seatNumber": "A-2",
                      "occupyToken": "token-2"
                    }
                  ]
                }
                """.formatted(concert.getConcertId());

        mockMvc.perform(post("/api/v1/tickets/reserve/schedule/{scheduleId}", schedule.getScheduleId())
                        .header("X-Queue-Token", "test-queue-token")
                        .with(user(securityUser))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(requestBody))
                .andDo(print())
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.resultCode").value("201-1"))
                .andExpect(jsonPath("$.msg").value("결제 및 티켓 생성 성공"))
                .andExpect(jsonPath("$.data").isArray())
                .andExpect(jsonPath("$.data.length()").value(2))

                .andExpect(jsonPath("$.data[0].ticketNumber").isString())
                .andExpect(jsonPath("$.data[0].urlPoster").value("poster.jpg"))
                .andExpect(jsonPath("$.data[0].concertName").value("싸이 콘서트"))
                .andExpect(jsonPath("$.data[0].seatNumber").value("A-1"))

                .andExpect(jsonPath("$.data[1].ticketNumber").isString())
                .andExpect(jsonPath("$.data[1].urlPoster").value("poster.jpg"))
                .andExpect(jsonPath("$.data[1].concertName").value("싸이 콘서트"))
                .andExpect(jsonPath("$.data[1].seatNumber").value("A-2"));

        assertThat(seat.getSeatStatus()).isEqualTo(SOLD_OUT);
        assertThat(ticketRepository.count()).isEqualTo(2);
    }

    @Test
    @DisplayName("티켓 4매 생성 실패")
    void createFourTickets() throws Exception {
        String requestBody = """
                {
                  "concertId": %d,
                  "seatHolds": [
                    { "seatNumber": "A-1", "occupyToken": "token-1" },
                    { "seatNumber": "A-2", "occupyToken": "token-2" },
                    { "seatNumber": "A-3", "occupyToken": "token-3" },
                    { "seatNumber": "A-4", "occupyToken": "token-4" }
                  ]
                }
                """.formatted(concert.getConcertId());

        mockMvc.perform(post("/api/v1/tickets/reserve/schedule/{scheduleId}", schedule.getScheduleId())
                        .header("X-Queue-Token", "test-queue-token")
                        .with(user(securityUser))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(requestBody))
                .andDo(print())
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.resultCode").value("400-2"))
                .andExpect(jsonPath("$.msg").value("회차당 최대 3매까지 예매 가능합니다."));
    }

    @Test
    @DisplayName("티켓 취소 성공")
    void cancelTicket() throws Exception {
        seat.updateSeatStatus(SOLD_OUT);
        Ticket ticket = ticketRepository.save(Ticket.create(user, schedule, seat, "ticket-number", seat.getSeatPrice()));

        mockMvc.perform(patch("/api/v1/tickets/cancel/{ticketId}", ticket.getTicketId())
                        .with(user(securityUser)))
                .andDo(print())
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.resultCode").value("200-1"))
                .andExpect(jsonPath("$.msg").value("티켓 취소 성공"));

        assertThat(ticket.isValid()).isFalse();
        assertThat(seat.getSeatStatus()).isEqualTo(AVAILABLE);
    }

    private User saveUser() {
        return userRepository.save(
                User.create(
                        "user1",
                        "user1@test.com",
                        "0000",
                        "테스트 유저",
                        LoginType.NORMAL
                )
        );
    }
}

File: back/src/test/kotlin/com/back/domain/ticket/controller/TicketControllerTest.kt
package com.back.domain.ticket.controller

import com.back.domain.concert.entity.Concert
import com.back.domain.concert.repository.ConcertRepository
import com.back.domain.schedule.constant.SeatStatus.HOLD
import com.back.domain.schedule.entity.Schedule
import com.back.domain.schedule.entity.ScheduleSeat
import com.back.domain.schedule.repository.ScheduleRepository
import com.back.domain.schedule.repository.ScheduleSeatRepository
import com.back.domain.ticket.entity.Ticket
import com.back.domain.ticket.repository.TicketRepository
import com.back.domain.user.constant.LoginType
import com.back.domain.user.entity.User
import com.back.domain.user.repository.UserRepository
import com.back.domain.venue.entity.Venue
import com.back.domain.venue.repository.VenueRepository
import com.back.global.security.SecurityUser
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.doAnswer
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc
import org.springframework.data.redis.core.HashOperations
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.data.redis.core.ValueOperations
import org.springframework.data.redis.core.ZSetOperations
import org.springframework.http.MediaType
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.user
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

@ActiveProfiles("test")
@SpringBootTest
@AutoConfigureMockMvc
@Transactional
class TicketControllerTest @Autowired constructor(
    private val mockMvc: MockMvc,
    private val userRepository: UserRepository,
    private val concertRepository: ConcertRepository,
    private val venueRepository: VenueRepository,
    private val scheduleRepository: ScheduleRepository,
    private val scheduleSeatRepository: ScheduleSeatRepository,
    private val ticketRepository: TicketRepository,
) {
    private lateinit var concert: Concert
    private lateinit var schedule: Schedule
    private lateinit var seat: ScheduleSeat
    private lateinit var userEntity: User
    private lateinit var securityUser: SecurityUser

    @MockitoBean
    private lateinit var stringRedisTemplate: StringRedisTemplate

    private lateinit var zSetOps: ZSetOperations<String, String>
    private lateinit var valOps: ValueOperations<String, String>
    private lateinit var hashOps: HashOperations<String, String, String>

    @BeforeEach
    @Suppress("UNCHECKED_CAST")
    fun setUp() {
        userEntity = saveUser()
        securityUser = SecurityUser(userEntity.userId!!, userEntity.name)

        zSetOps = mock(ZSetOperations::class.java) as ZSetOperations<String, String>
        valOps = mock(ValueOperations::class.java) as ValueOperations<String, String>
        hashOps = mock(HashOperations::class.java) as HashOperations<String, String, String>

        doReturn(zSetOps).`when`(stringRedisTemplate).opsForZSet()
        doReturn(valOps).`when`(stringRedisTemplate).opsForValue()
        doReturn(hashOps).`when`(stringRedisTemplate).opsForHash<String, String>()

        doReturn((System.currentTimeMillis() + 600000).toDouble()).`when`(zSetOps).score(anyString(), anyString())
        doReturn("test-queue-token").`when`(valOps).get(anyString())

        val defaultEntries = mapOf("userId" to userEntity.userId.toString(), "occupyToken" to "test-token")
        doReturn(defaultEntries).`when`(hashOps).entries(anyString())

        concert = concertRepository.save(
            Concert.create("싸이 콘서트", "설명", LocalDateTime.now(), LocalDateTime.now().plusDays(1), "poster.jpg")
        )
        val venue = venueRepository.save(Venue.create("공연장", "서울", 15000L))
        schedule = scheduleRepository.save(Schedule.create(concert, venue, LocalDateTime.now().plusHours(12), 1))

        seat = scheduleSeatRepository.save(ScheduleSeat.create(schedule, "VIP", "A-1", 150000, HOLD))
        scheduleSeatRepository.save(ScheduleSeat.create(schedule, "VIP", "A-2", 150000, HOLD))
        scheduleSeatRepository.save(ScheduleSeat.create(schedule, "VIP", "A-3", 150000, HOLD))
        scheduleSeatRepository.save(ScheduleSeat.create(schedule, "VIP", "A-4", 150000, HOLD))
    }

    @Test
    @DisplayName("티켓 2매 생성 성공")
    @Suppress("UNCHECKED_CAST")
    fun createTicket() {
        val mapAnswer = { invocation: org.mockito.invocation.InvocationOnMock ->
            val key = invocation.getArgument(0, String::class.java)
            when {
                key.endsWith("A-1") -> mapOf("userId" to userEntity.userId.toString(), "occupyToken" to "token-1")
                key.endsWith("A-2") -> mapOf("userId" to userEntity.userId.toString(), "occupyToken" to "token-2")
                else -> mapOf("userId" to userEntity.userId.toString(), "occupyToken" to "test-token")
            }
        }
        doAnswer(mapAnswer).`when`(hashOps).entries(anyString())

        val requestBody = """
            {
              "concertId": ${concert.concertId},
              "seatHolds": [
                {
                  "seatNumber": "A-1",
                  "occupyToken": "token-1"
                },
                {
                  "seatNumber": "A-2",
                  "occupyToken": "token-2"
                }
              ]
            }
        """.trimIndent()

        mockMvc.perform(
            post("/api/v1/tickets/reserve/schedule/{scheduleId}", schedule.scheduleId)
                .header("X-Queue-Token", "test-queue-token")
                .with(user(securityUser))
                .contentType(MediaType.APPLICATION_JSON)
                .content(requestBody)
        )
            .andDo(print())
            .andExpect(status().isCreated)
            .andExpect(jsonPath("$.resultCode").value("201-1"))
            .andExpect(jsonPath("$.data.length()").value(2))
    }

    @Test
    @DisplayName("티켓 취소 성공")
    fun cancelTicket() {
        val savedTicket = ticketRepository.save(
            Ticket.create(userEntity, schedule, seat, "TICKET-12345", 150000, "GROUP-1")
        )

        mockMvc.perform(
            patch("/api/v1/tickets/cancel/{ticketId}", savedTicket.ticketId)
                .with(user(securityUser))
        )
            .andDo(print())
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.resultCode").value("200-1"))

        val updatedTicket = ticketRepository.findById(savedTicket.ticketId!!).get()
        assertThat(updatedTicket.isValid).isFalse()
    }

    private fun saveUser(): User {
        return userRepository.save(
            User.create("user1", "user@test.com", "password", "테스트유저", LoginType.NORMAL)
        )
    }
}
